pangeo_forge_recipes/recipes/reference_hdf_zarr.py

```diff
@@ -0,0 +1,202 @@
+import functools
+import json
+import os
+import tempfile
+from dataclasses import dataclass, field
+from typing import Callable, Hashable, Iterable, List, Optional, Union
+
+import fsspec
+import yaml
+from fsspec_reference_maker.combine import MultiZarrToZarr
+from fsspec_reference_maker.hdf import SingleHdf5ToZarr
+
+from .base import BaseRecipe
+
+
+def no_op(*_, **__) -> None:
+    """A function that does nothing, regardless of inputs"""
+    return None
+
+
+@dataclass
+class ReferenceHDFRecipe(BaseRecipe):
+    """
+    Generates reference files for each input netCDF, then combines
+    into one ensemble output
+
+    Currently supports concat or merge along a single dimension.
+    TODO: multiple could be made by calling MultiZarrToZarr once
+     for each dimension
+
+    See fsspec-reference-maker and fsspec's ReferenceFileSystem
+    To use this class, you must have fsspec-reference-maker, ujson,
+    xarray, fsspec, zarr, h5py and ujson in your recipe's requirements.
+
+    This class will also produce an Intake catalog stub in YAML
+    format, with the same name as the output json + ".yaml". You
+    can use intake (and intake-xarray) to load the dataset, and this
+    is the recommended way to distribute access.
+
+    SingleHdf5ToZarr parameters:
+
+    :param netcdf_url: location of the original data files. Can be a
+        string (will be glob-expanded) or explicit list of paths or
+        list of glob-strings. May eventually support FilePatterns.
+        Should include protocol specifier, as usually formulated for
+        fsspec.
+    :param netcdf_storage_options: dict of kwargs for creating fsspec
+        instance to read original data files
+    :param inline_threshold: blocks with fewer bytes than this will be
+        inlined into the output reference file
+
+    MultiZarrToZarr parameters (if accessing more than one HDF):
+
+    :param xarray_open_kwargs: kwargs passed to xarray.open_dataset
+    :param xarray_concat_args: kwargs passed to xarray.concat
+
+    Output parameters:
+
+    :param output_url: where the final reference file is written.
+    :param output_storage_options: dict of kwargs for creating fsspec
+        instance when writing final output
+    :param template_count: the number of occurrences of a URL before it
+        gets made a template. ``None`` to disable templating
+    """
+
+    # TODO: support chunked ("tree") aggregation: would entail processing each file
+    #  in one stage, running a set of merges in a second step and doing
+    #  a master merge in finalise. This would maybe map to iter_chunk,
+    #  store_chunk, finalize_target. The strategy was used for NWM's 370k files.
+
+    # TODO: as written, this rcipe is specific to HDF5 files. fsspec-reference-maker
+    #  also supports TIFF and grib2 (and more coming)
+    netcdf_url: Union[str, List[str]]
+    output_url: str
+    _work_dir: str = ""
```

Comment:
We should use a `MetadataTarget` or something similar from `strorage.py` here instead.

Reply:
In this recipe framework, it's only meant to be a temporary storage, and it's only the global aggregated JSON file that is output. Where do temporary things go?

Reply:
CacheTarget (for arbitrary binary data) or MetadataTarget for temporary metadata. Here is an example of setting

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c4867719c4866ec46436b5d53d95ceb470a6f72e/pangeo_forge_recipes/recipes/xarray_zarr.py#L143-L144

and getting

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c4867719c4866ec46436b5d53d95ceb470a6f72e/pangeo_forge_recipes/recipes/xarray_zarr.py#L395

metadata from the cache.